packages/vite/src/node/server/pluginContainer.ts

```diff
@@ -571,12 +584,27 @@ export async function createPluginContainer(
   }
 
   let closed = false
+  const processesing = new Set<Promise<any>>()
+  function awaitPromiseOnClose<T>(promise: Promise<T>) {
+    processesing.add(promise)
+    return promise.finally(() => processesing.delete(promise))
+  }
+  function handleHookPromise<T>(maybePromise: undefined | T | Promise<T>) {
```

Comment:
could be helpful to add a comment
```suggestion
  /** keeps track of hook promises so that we can wait for them all to finish upon closing the server */
  function handleHookPromise<T>(maybePromise: undefined | T | Promise<T>) {
```